parlai/agents/transformer/modules.py

```diff
@@ -1039,19 +1109,21 @@ def reorder_encoder_states(self, encoder_states, indices):
         return enc, mask
 
     def reorder_decoder_incremental_state(
-        self, incremental_state: Dict[int, dict], inds: torch.Tensor
-    ) -> Dict[int, dict]:
+        self, incremental_state: Dict[str, torch.Tensor], inds: torch.LongTensor
+    ) -> Dict[str, torch.Tensor]:
         """
         Reorder the decoder incremental state.
 
         See ``TorchGeneratorModel.reorder_decoder_incremental_state`` for a description.
 
-        Here, incremental_state is a dict whose keys are layer indices and whose values
-        are dicts containing the incremental state for that layer.
+        Here, incremental_state is a dict whose keys are attention types (self- or
+        enc/dec and prev key, prev value, or prev mask) and whose values are Tensors,
+        with matrices for each layer concatenated in the last dimension. The 0th
+        dimension is the batch size and thus gets reindexed.
         """
         return {
-            idx: layer.reorder_incremental_state(incremental_state[idx], inds)
-            for idx, layer in enumerate(self.decoder.layers)
+            key: torch.index_select(val, 0, inds.to(val.device)).contiguous()
```

Comment:
I personally felt like the previous method was much cleaner. Why can't we torchscript those things?

Reply:
TorchScript wasn't able to handle the original Dict[Dict[Dict[str,Tensor]]] data structure